### PR TITLE
Some debug-span and debug-log changes to help with filtering during tracing analysis

### PR DESCRIPTION
**Pull request by tayfunelmas**
This is not the complete list of changes I plan to do but wanted to get a first batch reviewed first.

There are two kinds of changes applied to debug logs and traces that would help filtering the them when analyzing traces (either through the new tracing UI or through the log files):

1) Simplify the name, eg. instead of full sentence (unless sentence makes more sense), replace with an identifier-like name (in most cases reflecting the type of operation or function). Move the params substituted in the string to separate log/span fields. 

2) Add missing fields and attempt to standardize the naming for the fields so that when it comes to filtering traces for a certain analysis we know what kind of fields are available across the logs/traces for the same kind of entity/operation. For example:
- `shard_id` for shard id (also prefer shard_id over UID as it has extra version and it can be obtain by other means).
- `sync_hash` for hash of the state-sync
- `height` for the block height (use last_has and prev_hash if it is not clear from the context)
- `block_hash` for hash of the block
- `chunk_hash/chunk_hashes` for hash of the chunk
- `part_id` for the id of the state-sync part
- `error` for the error type or message
- `sync_type` for the sync type (eg. block, head, state)
- `height_included` for the block height a chunk is included
- `height_created` for the block height a chunk is created

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -1080,7 +1080,7 @@ impl Chain {
         header: &BlockHeader,
         challenges: &mut Vec<ChallengeBody>,
     ) -> Result<(), Error> {
-        debug!(target: "chain", "Process block header: {} at {}", header.hash(), header.height());
+        debug!(target: "chain", block_hash=?header.hash(), height=header.height(), "process_block_header", );
```
Comment (nagisa) [inline chain/chain/src/chain.rs:1083]:
```suggestion
        debug!(target: "chain", block_hash=?header.hash(), height=header.height(), "process_block_header");
```

Unfortunately rustfmt can't really format some macro invocations, so we have to do it by hand... T_T

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -1920,12 +1919,7 @@ impl Chain {
                     true,
                 )
             };
-            tracing::debug!(
-                target: "chain",
-                "Updating flat storage for shard {} need_flat_storage_update: {}",
-                shard_id,
-                need_flat_storage_update
-            );
+            tracing::debug!(target: "chain",shard_id,need_flat_storage_update,"Updating flat storage");
```
Comment (nagisa) [inline chain/chain/src/chain.rs:1922]:
```suggestion
            tracing::debug!(target: "chain", shard_id, need_flat_storage_update, "Updating flat storage");
```

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -1750,10 +1750,9 @@ impl Chain {
                     Error::BlockKnown(block_known_error) => {
                         debug!(
                             target: "chain",
-                            "Block {} at {} is known at this time: {:?}",
-                            block.hash(),
-                            block_height,
-                            block_known_error);
+                            block_hash=?block.hash(),
+                            height=block_height,
+                            error=?block_known_error, "Block known at this time");
```
Comment (nagisa) [inline chain/chain/src/chain.rs:1755]:
```suggestion
                            error=?block_known_error,
                            "Block known at this time");
```

File: chain/chain/src/chain.rs [hunk 4]
```diff
@@ -1738,8 +1739,7 @@ impl Chain {
                             .add_block_with_missing_chunks(orphan, missing_chunk_hashes.clone());
                         debug!(
                             target: "chain",
-                            "Process block: missing chunks. Block hash: {:?}. Missing chunks: {:?}",
-                            block_hash, missing_chunk_hashes,
+                            ?block_hash, chunk_hashes=missing_chunk_hashes.iter().map(|h| format!("{:?}", h)).join(","), "Process block: missing chunks"
```
Comment (nagisa) [inline chain/chain/src/chain.rs:1742]:
```suggestion
                            ?block_hash,
                            chunk_hashes=missing_chunk_hashes.iter().map(|h| format!("{:?}", h)).join(","),
                            "Process block: missing chunks"
```

File: chain/chain/src/chain.rs [hunk 5]
```diff
@@ -2525,7 +2519,7 @@ impl Chain {
             "get_state_response_part",
             shard_id,
             part_id,
-            %sync_hash)
+            ?sync_hash)
```
Comment (nagisa) [inline chain/chain/src/chain.rs:2522]: Why was this change necessary? By default Display is intended for anything reaches user eyes, including logs...

File: chain/chain/src/chain_update.rs [hunk 6]
```diff
@@ -171,7 +171,7 @@ impl<'a> ChainUpdate<'a> {
         let height = block.header().height();
         match resharding_results {
             ReshardingResults::ApplyReshardingResults(mut results) => {
-                tracing::debug!(target: "resharding", height, ?shard_uid, "process_resharding_results apply");
+                tracing::debug!(target: "resharding", height, shard_id=?shard_uid.shard_id(), "process_resharding_results apply");
```
Comment (nagisa) [inline chain/chain/src/chain_update.rs:174]: Why was this change necessary? I thought we are prioritizing use of `shard_uid` to avoid ambiguities?

File: chain/client/src/client.rs [hunk 7]
```diff
@@ -852,6 +845,32 @@ impl Client {
...
+        epoch_id: &EpochId,
+        last_header: ShardChunkHeader,
+        next_height: BlockHeight,
+        shard_id: ShardId,
+        validator_signer: Arc<dyn ValidatorSigner>,
+    ) -> Result<Option<ProduceChunkResult>, Error> {
+        let timer = Instant::now();
+        let _timer =
+            metrics::PRODUCE_CHUNK_TIME.with_label_values(&[&shard_id.to_string()]).start_timer();
+        let span = tracing::debug_span!(target: "client", "produce_chunk", height=next_height, shard_id, ?epoch_id).entered();
```
Comment (nagisa) [inline chain/client/src/client.rs:871]: FWIW I also recommend moving to the `#[instrument]` macro for spans that cover roughly the entire function. But its a nitpick more so than anything else.

File: chain/client/src/client.rs [hunk 8]
```diff
@@ -914,6 +933,7 @@ impl Client {
             protocol_version,
         )?;
 
+        span.record("chunk_hash", format!("{:?}", encoded_chunk.chunk_hash()));
```
Comment (nagisa) [inline chain/client/src/client.rs:936]: Avoid using `format` here. It will force formatting of the value regardless of whether the span is being emit or not. Use https://docs.rs/tracing/latest/tracing/field/fn.debug.html instead.

File: chain/client/src/stateless_validation/chunk_endorsement_tracker.rs [hunk 9]
```diff
@@ -87,10 +87,10 @@ impl ChunkEndorsementTracker {
         let Some(chunk_endorsements) = chunk_endorsements else {
             return;
         };
-        tracing::debug!(target: "stateless_validation", ?chunk_hash, "Processing pending chunk endorsements.");
+        tracing::debug!(target: "client", ?chunk_hash, "Processing pending chunk endorsements.");
         for endorsement in chunk_endorsements.values() {
             if let Err(error) = self.process_chunk_endorsement(chunk_header, endorsement.clone()) {
-                tracing::debug!(target: "stateless_validation", ?endorsement, "Error processing pending chunk endorsement: {:?}", error);
+                tracing::debug!(target: "client", ?endorsement, "Error processing pending chunk endorsement: {:?}", error);
```
Comment (nagisa) [inline chain/client/src/stateless_validation/chunk_endorsement_tracker.rs:93]:
```suggestion
                tracing::debug!(target: "client", ?endorsement, ?error, "Error processing pending chunk endorsement");
```

File: chain/chain/src/chain.rs (hunk 5 above)
Comment (robin-near) [reply chain/chain/src/chain.rs:2522]: Looks like they are the same

```
impl fmt::Debug for CryptoHash {
    fn fmt(&self, fmtr: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self, fmtr)
    }
}

impl fmt::Display for CryptoHash {
    fn fmt(&self, fmtr: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.to_base58_impl(|encoded| fmtr.write_str(encoded))
    }
}
```

(Though for hashes in particular, I have a habit of using Debug, because I think the ethereum libraries implement Display by displaying the hash with ellipsis in the middle and it's really annoying. I guess Near code doesn't do that but the habit and fear has already formed :) )

File: chain/chain/src/chain_update.rs (hunk 6 above)
Comment (robin-near) [reply chain/chain/src/chain_update.rs:174]: For resharding we should indeed use shard_uid. I'm guessing Tayfun wants the shard_id for consistent analysis, but we can export both fields.

File: chain/client/src/client.rs (hunk 8 above)
Comment (robin-near) [reply chain/client/src/client.rs:936]: This is a very nice addition to this span btw, I didn't know you could do this!

File: chain/client/src/stateless_validation/state_witness_producer.rs [hunk 10]
```diff
@@ -34,6 +34,9 @@ impl Client {
         if !checked_feature!("stable", StatelessValidationV0, protocol_version) {
             return Ok(());
         }
+        let chunk_header = chunk.cloned_header();
+        let shard_id = chunk_header.shard_id();
```
Comment (robin-near) [inline chain/client/src/stateless_validation/state_witness_producer.rs:38]: did you mean to use the shard_id as a the span field?

**Review by robin-near — APPROVED** [review]
What's the significance of the "target" parameter in the tracing analysis? How should we choose the target going forward?

**Comment by nagisa** [comment]
> What's the significance of the "target" parameter in the tracing analysis?

It aids the ability to filter (out) the (un)interesting parts of the code. I personally think that a single word that's used in too many spans/events is insufficiently precise, but this is something that each of us needs to discover independently, unfortunately.

File: chain/chain/src/chain_update.rs (hunk 6 above)
Comment (nagisa) [reply chain/chain/src/chain_update.rs:174]: > we can export both fields

Conversely traces are measured by byte. Its fine to do this if the event is rare enough, but for frequently triggered output we should stay mindful of not duplicating information unnecessarily.

**Comment by robin-near** [comment]
@nagisa I understand that it aids filtering at the logging level, but this is talking about tracing where we have the ability to take in all traces and then filter/analyze them later. So I wanted to know how @tayfunelmas is using the target tag in his analysis, and if anyone changes the tags in the future, what impact it would have on the tracing analysis tooling.

**Comment by nagisa** [comment]
> but this is talking about tracing where we have the ability to take in all traces and then filter/analyze them later.

It isn't actually practical to send *everything* that's traced off to somewhere in many cases. It is already a constant >500KiB/s of ingest with the few things we trace at the debug level and that's already enough for grafana tempo to start dropping some of the ingest traffic at our fairly conservative default ingest limits. And with an increase of those limits, so would increase our monthly cost (this holds true regardless of where we ship the traces off to, unless they are held locally for immediate inspection, such as with your tool; but even then storage ain't free…)

If we didn't filter our traces at the emitter we'd be looking at potentially dozens of megabytes of traces per second. Components like the compiler or hyper are particularly chatty. So even for traces it is important to have well thought-out targets and levels. Otherwise even just gathering an useful trace becomes a chore (as we recently found out -- enabling host function tracing ended up with traces so truncated that they were largely incomprehensible…)

File: chain/chain/src/chain.rs (hunk 5 above)
Comment (tayfunelmas) [reply chain/chain/src/chain.rs:2525]: Also for the consistency because in majority of other places hashes are logged with Debug.

File: chain/chain/src/chain_update.rs (hunk 6 above)
Comment (tayfunelmas) [reply chain/chain/src/chain_update.rs:174]: Ok, let me revert that back to shard uid and we can handle it in the tracing collector/analyzer side. no need to add extra field with the same info.

File: chain/client/src/client.rs (hunk 8 above)
Comment (tayfunelmas) [reply chain/client/src/client.rs:936]: done.

**Comment by tayfunelmas** [comment]
> What's the significance of the "target" parameter in the tracing analysis? How should we choose the target going forward?

Actually both useful and not. First, I am planning to primarily use the span/log names and the fields. The target field will only help to filter down the logs, even though the data that needs to be processed may be still large, but better than having everything. I am expecting that for certain analysis (eg. chunk/block production), the interesting pieces of the logs will be coming from certain targets such as "client" and "network", this is why I did some changes to the target names for the events that I think are interesting.

File: chain/client/src/stateless_validation/state_witness_producer.rs (hunk 10 above)
Comment (tayfunelmas) [reply chain/client/src/stateless_validation/state_witness_producer.rs:38]: yes. done.

File: chain/client/src/client.rs (hunk 7 above)
Comment (tayfunelmas) [reply chain/client/src/client.rs]: done.

**Comment by nagisa** [comment]
> How should we choose the target going forward?

Our style guide, by the way, has some guidance on that, it reads as such:

> Always specify the target explicitly. A good default value to use is the crate name, or the module path (e.g. chain::client) so that events and spans common to a topic can be grouped together. This grouping can later be used for customizing which events to output.
